### ISSUE TITLE
Prepending a path to an existing endpoint

### ISSUE THREAD

**Issue by valenterry** [softwaremill/tapir#148]
I'm trying to prepend a fixed path to an existing endpoint. E.g. I have a few endpoints and I have a versioned API so I'd like to do something like `endpoints.map(_.prependPath("api/v1/")`.

I tried to do it with a `ServerEndpoint` like that:
```scala
val prefixPath: EndpointInput.FixedPath = ???
serverEndpoint.copy(
  endpoint = serverEndpoint.endpoint.copy(
    input = prefixPath / serverEndpoint.endpoint.input
  )
)	
```
It compiles but then I only get 404s back - probably connecting the prefix to the existing input does not work like it.

Any ideas?

**Comment by valenterry** [comment]
Apparently the problem occurred because I used `EndpointInput.FixedPath("api/v1")` which probably escaped the slash.

**Comment by adamw** [comment]
Ah yes, the `/` inside will be escaped as it's then treated as a single path component.
Prepending inputs can be made more general, as we have now e.g. `endpoint.in(...)` - this appends an input, similarly `endpoint.prependIn(...)` could prepend the input.

**Comment by ghostbuster91** [comment]
Is it valuable to pollute api with additional method for something which can be achieved in other way (building upon base endpoint which has "api/v1")?

**Comment by valenterry** [comment]
> Is it valuable to pollute api with additional method for something which can be achieved in other way (building upon base endpoint which has "api/v1")?

The alternative is that any common "configuration" (such as path-prefix) needs to be passed all the way down to the point where the other endpoints are created. This is the opposite of compositibility. 

Just imagine there is a library that provides me with an endpoint that I want to prefix it, but the library author didn't not anticipate that I might want to do something. If he provides me for example some http4s `HttpRoutes`, I can take them and prepend. With tapir, currently I can do it also (in the way I described) but not in they way you suggest here (building upon a base endpoint).

**Comment by ghostbuster91** [comment]
Ok, that is a fair point, let's do it :)

**Comment by adamw** [comment]
Released in 0.9.3
